device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/auth/IotHubSasTokenProvidedAuthenticationProvider.java

```diff
@@ -0,0 +1,68 @@
+/*
+*  Copyright (c) Microsoft. All rights reserved.
+*  Licensed under the MIT license. See LICENSE file in the project root for full license information.
+*/
+
+package com.microsoft.azure.sdk.iot.device.auth;
+
+import com.microsoft.azure.sdk.iot.device.SasTokenProvider;
+
+import javax.net.ssl.SSLContext;
+
+public class IotHubSasTokenProvidedAuthenticationProvider extends IotHubSasTokenAuthenticationProvider
+{
+    SasTokenProvider sasTokenProvider;
+    char[] lastSasToken;
+
+    public IotHubSasTokenProvidedAuthenticationProvider(String hostName, String deviceId, String moduleId, SasTokenProvider sasTokenProvider, SSLContext sslContext) {
+        super(hostName, null, deviceId, moduleId, sslContext);
+
+        if (sasTokenProvider == null)
+        {
+            throw new IllegalArgumentException("sas token provider cannot be null");
+        }
+
+        this.sasTokenProvider = sasTokenProvider;
+    }
+
+    @Override
+    public boolean isRenewalNecessary()
+    {
+        return false;
```

Comment:
Why is this `false`? If it can be refreshed, wouldn't renewal be necessary?

Reply:
This function is more about "Does the customer need to create a new client because they created this client with a SAS token rather than a symmetric key, and that SAS token has expired"

Reply:
Since the user is providing all the SAS tokens, this function should never return true.

Reply:
Or maybe `canRefreshToken()` should be false then?

Reply:
Oh, this isn't the impl for when a SAS-token provided? I'm confused.

Reply:
Let me just rename this method to help with the confusion.